src/Storages/StorageAggregatingMemory.cpp

```diff
@@ -0,0 +1,513 @@
+                              settings.group_by_two_level_threshold,
+                              settings.group_by_two_level_threshold_bytes,
+                              settings.max_bytes_before_external_group_by,
+                              settings.empty_result_for_aggregation_by_empty_set,
+                              select_context->getTemporaryVolume(),
+                              settings.max_threads,
+                              settings.min_free_disk_space_for_temporary_data,
+                              true);
+
+    aggregator_transform = std::make_shared<AggregatingTransformParams>(params, false);
+
+    initState(constructor_context);
+
+    is_initialized = true;
+}
+
+void StorageAggregatingMemory::initState(ContextPtr context)
+{
+    many_data = std::make_shared<ManyAggregatedData>(1);
+
+    /// If there was no data, and we aggregate without keys, and we must return single row with the result of empty aggregation.
+    /// To do this, we pass a block with zero rows to aggregate.
+    if (aggregator_transform->params.keys_size == 0 && !aggregator_transform->params.empty_result_for_aggregation_by_empty_set)
+    {
+        AggregatingOutputStream os(*this, getInMemoryMetadataPtr(), context);
+        os.write(src_block_header);
+    }
+}
+
+void StorageAggregatingMemory::startup()
+{
+    try
+    {
+        lazyInit();
+    }
+    catch (Exception & e)
+    {
+        e.addMessage("Failed to initialize AggregatingMemory on startup");
+        LOG_ERROR(log, "{}", getCurrentExceptionMessage(true));
+    }
+}
+
+Pipe StorageAggregatingMemory::read(
+    const Names & column_names,
+    const StorageMetadataPtr & metadata_snapshot,
+    SelectQueryInfo & query_info,
+    ContextPtr context,
+    QueryProcessingStage::Enum /*processed_stage*/,
+    size_t /*max_block_size*/,
+    unsigned num_streams)
+{
+    lazyInit();
+    metadata_snapshot->check(column_names, getVirtuals(), getStorageID());
+    // TODO implement O(1) read by aggregation key
+
+    auto prepared_data = aggregator_transform->aggregator.prepareVariantsToMerge(many_data->variants);
+    auto prepared_data_ptr = std::make_shared<ManyAggregatedDataVariants>(std::move(prepared_data));
+
+    ProcessorPtr source;
+
+    auto filter_key = getFilterKeys(aggregator_transform->params, query_info);
+    source = std::make_shared<ConvertingAggregatedToChunksTransform>(aggregator_transform, std::move(prepared_data_ptr), num_streams);
+
+    StoragePtr mergable_storage = StorageSource::create(source_storage->getStorageID(), source_storage->getInMemoryMetadataPtr()->getColumns(), source);
+
+    ContextMutablePtr local_context = Context::createCopy(context);
+    local_context->addViewSource(mergable_storage);
+
+    InterpreterSelectQuery select(metadata_snapshot->getSelectQuery().inner_query, local_context, SelectQueryOptions());
+    BlockIO select_result = select.execute();
+
+    return QueryPipeline::getPipe(std::move(select_result.pipeline));
```

Comment:
Looks fine.
However, looks like there is no synchronization at all.

Reply:
We should decide how to protect `many_data->variants`.
We can just use mutex. It is ok, but reads and writes will block each other. Code will be like single-threaded.
Also, we can use RW lock. But slow insert may make table unusable.

I think, it would be better to make a deep copy of `many_data->variants` for every insert (BTW, we can't just copy every arena there, cause aggregate states are not POD. But we can recreate every state in new memory and merge them). We still could not do more then one insert a time, but selects could use the last available version of `many_data->variants` and not block each other or selects.